src/metabase/related.clj

```diff
@@ -0,0 +1,249 @@
+             [collection :refer [Collection]]
+             [dashboard :refer [Dashboard]]
+             [dashboard-card :refer [DashboardCard]]
+             [field :refer [Field]]
+             [interface :as mi]
+             [metric :refer [Metric]]
+             [query :refer [Query]]
+             [segment :refer [Segment]]
+             [table :refer [Table]]]
+            [schema.core :as s]
+            [toucan.db :as db]))
+
+(def ^:private ^Long max-best-matches        3)
+(def ^:private ^Long max-serendipity-matches 2)
+(def ^:private ^Long max-matches             (+ max-best-matches
+                                                max-serendipity-matches))
+
+(def ^:private ContextBearingForm
+  [(s/one (s/constrained (s/either s/Str s/Keyword)
+                         (comp #{"field-id" "metric" "segment" "fk->"}
+                               str/lower-case
+                               name))
+          "head")
+   s/Any])
+
+(defn- collect-context-bearing-forms
+  [form]
+  (into #{}
+    (remove (s/checker ContextBearingForm))
+    (tree-seq sequential? identity form)))
+
+(defmulti
+  ^{:doc "Return the relevant parts of a given entity's definition.
+          Relevant parts are those that carry semantic meaning, and especially
+          context-bearing forms."
+    :arglists '([entity])}
+  definition type)
+
+(defmethod definition (type Card)
+  [card]
+  (-> card
+      :dataset_query
+      :query
+      ((juxt :breakout :aggregation :expressions :fields))))
+
+(defmethod definition (type Metric)
+  [metric]
+  (-> metric :definition ((juxt :aggregation :filter))))
+
+(defmethod definition (type Segment)
+  [segment]
+  (-> segment :definition :filter))
+
+(defn similarity
+  "How similar are entities `a` and `b` based on a structural comparison of their
+   definition (MBQL).
+   For the purposes of finding related entites we are only interested in
+   context-bearing subforms (field, segment, and metric references). We also
+   don't care about generalizations (less context-bearing forms) and refinements
+   (more context-bearing forms), so we just check if the less specifc form is a
+   subset of the more specific one."
+  [a b]
+  (let [context-a (-> a definition collect-context-bearing-forms)
+        context-b (-> b definition collect-context-bearing-forms)]
+    (/ (count (set/intersection context-a context-b))
+       (max (min (count context-a) (count context-b)) 1))))
+
+(defn- rank-by-similarity
+  [reference entities]
+  (->> entities
+       (remove #{reference})
+       (map #(assoc % :similarity (similarity reference %)))
+       (sort-by :similarity >)))
+
+(defn- interesting-mix
+  "Create an interesting mix of matches. The idea is to have a balanced mix
+   between close (best) matches and more diverse matches to cover a wider field
+   of intents."
+  [matches]
+  (let [[best rest] (split-at max-best-matches matches)]
+    (concat best (->> rest shuffle (take max-serendipity-matches)))))
```

Comment:
I like this for v1

Reply:
(missing) permission checking added. As well as checking for archived/hidden.